### ISSUE TITLE
"Invalid owner name: empy label", "syntax error" from acmeproxy.pl when using dns_nsupdate

### ISSUE THREAD

**Issue by BAM279** [MadCamel/acmeproxy.pl#3]
Hi,

first of all, thank you very much for creating acmeproxy.pl.

I have done the initial acmeproxy.pl config using the dns_nsupdate dns api against my local bind dns server and acmeproxy has successfully updated the dns txt record of the zone using my configured tsig key and provisioned its own acmeproxy ssl cert from my local step-ca internal acme server.

Acmeproxy.pl mojolicious is up and running on its static ip with its new ssl cert and I can browse to the web page and confirm all is good.

The environment setup appears valid and ready to run an actual acme.sh request against the acmeproxy service basically.

When I do so however, the acmeproxy.pl output reports an error whilst trying to run the nsupdate to add the _acme-challenge dns txt record:

Acmeproxy.pl output:

[2024-08-07 19:45:27.71508] [5410] [info] Listening at "https://*:9096?cert=%2Froot%2F.acme.sh%2Facmeproxy.pl.crt&key=%2Froot%2F.acme.sh%2Facmeproxy.pl.key"
[2024-08-07 19:49:06.42301] [5410] [trace] [YBw8BWNgrRuh] POST "/present"
[2024-08-07 19:49:06.42341] [5410] [trace] [YBw8BWNgrRuh] Routing to a callback
Web application available at https://127.0.0.1:9096
[Wed Aug  7 07:49:06 PM UTC 2024] HTTP: 10.2.53.3 POST https://acmeproxy.2.internal.lan:9096/present
[Wed Aug  7 07:49:06 PM UTC 2024] auth: acmeuser successfully authenticated for _acme-challenge.pihole2.2.internal.lan.
[Wed Aug  7 07:49:06 PM UTC 2024] executing: /usr/bin/env bash -c "source /root/.acme.sh/acme.sh >/dev/null 2>&1; source /root/.acme.sh/dnsapi/dns_nsupdate.sh; dns_nsupdate_rm \"_acme-challenge.pihole2.2.internal.lan.\" \"SC4z5IM2MHMBVzYjCEWh9Bt6iYtwbhsJqiD9d0TrDJ4\";"
[Wed Aug  7 19:49:06 UTC 2024] removing _acme-challenge.pihole2.2.internal.lan.. txt
invalid owner name: empty label
syntax error
[Wed Aug  7 19:49:06 UTC 2024] error updating domain
[Wed Aug  7 07:49:06 PM UTC 2024] executing: /usr/bin/env bash -c "source /root/.acme.sh/acme.sh >/dev/null 2>&1; source /root/.acme.sh/dnsapi/dns_nsupdate.sh; dns_nsupdate_add \"_acme-challenge.pihole2.2.internal.lan.\" \"SC4z5IM2MHMBVzYjCEWh9Bt6iYtwbhsJqiD9d0TrDJ4\";"
[Wed Aug  7 19:49:06 UTC 2024] adding _acme-challenge.pihole2.2.internal.lan.. 60 in txt "SC4z5IM2MHMBVzYjCEWh9Bt6iYtwbhsJqiD9d0TrDJ4"
invalid owner name: empty label
syntax error
[Wed Aug  7 19:49:06 UTC 2024] error updating domain
[2024-08-07 19:49:06.57932] [5410] [trace] [YBw8BWNgrRuh] 200 OK (0.15628s, 6.399/s)

No nsupdate request appears to actually get transmitted on the network, and on the bind dns side, the only dns queries received from the client are a basic txt record lookup for the _acme-challenge entry, which were never added.

Im a bit confused as the nsupdate worked fine when acmeproxy got its own cert for its own web service, yet is unable to do the exact same procedure when a request is recieved from another client via the acmeproxy web service.  The logs confirm the authentication is good, which seems to indicate it fails at passing along the request details to the nsupdate piece.

The only thing I notice in the acmeproxy.pl output, is that the request fdqn of the cert has a double-period at the end during the record "adding" and "removing" stages in the log output.  I am not sure if this is of significance or is merely cosmetic but it is all I can come up with.

Any advice or suggestions would be greatly appreciated.

Best regards,
BAM

**Comment by MadCamel** [comment]
You're welcome, I'm happy when people find things I've made to be useful. Excellent use-case as well, I've never thought of using this with an internal CA and acme server!

That additional . at the end is indeed suspicious, I think your spidey sense is on point. I'm not in a position to replicate your entire setup to figure out how it's creeping into the hostname so let's just remove it before calling the acme.sh routines.

In line 111 of acmeproxy.pl make it look like this:
```perl
sub acme_cmd ($action, $fqdn, $value) {
  $fqdn =~ s/\.+$/./; # Add this line
# Let's not pass weird characters to a shell
  return "invalid characters in fqdn" unless ($fqdn =~ /^[\w_\.-]+$/);
...
```

If this works please let me know and I'll add a slightly cleaner fix to the code

**Comment by MadCamel** [comment]
Thanks for trying. Try this instead:
```perl
$fqdn =~ s/\.+$//;
```

**Comment by BAM279** [comment]
Hey, that has resolved it... 

Acmeproxy.pl output:

root@acmeproxy:~# ./acmeproxy.pl 
[2024-08-09 12:48:07.58230] [6344] [info] Listening at "https://*:9096?cert=%2Froot%2F.acme.sh%2Facmeproxy.pl.crt&key=%2Froot%2F.acme.sh%2Facmeproxy.pl.key"
Web application available at https://127.0.0.1:9096
[Fri Aug  9 12:49:32 PM UTC 2024] HTTP: 10.2.53.3 POST https://acmeproxy.2.internal.lan:9096/present
[2024-08-09 12:49:32.51832] [6344] [trace] [bnfHb5pCT8H9] POST "/present"
[2024-08-09 12:49:32.51872] [6344] [trace] [bnfHb5pCT8H9] Routing to a callback
[Fri Aug  9 12:49:32 PM UTC 2024] auth: acmeuser successfully authenticated for _acme-challenge.pihole2.2.internal.lan.
[Fri Aug  9 12:49:32 PM UTC 2024] executing: /usr/bin/env bash -c "source /root/.acme.sh/acme.sh >/dev/null 2>&1; source /root/.acme.sh/dnsapi/dns_nsupdate.sh; dns_nsupdate_rm \"_acme-challenge.pihole2.2.internal.lan\" \"AVbCsG7_AB0lAzIQDxLBj0c9-gVEntFkaZkfmeoxtLA\";"
[Fri Aug  9 12:49:32 UTC 2024] removing _acme-challenge.pihole2.2.internal.lan. txt
[Fri Aug  9 12:49:32 PM UTC 2024] executing: /usr/bin/env bash -c "source /root/.acme.sh/acme.sh >/dev/null 2>&1; source /root/.acme.sh/dnsapi/dns_nsupdate.sh; dns_nsupdate_add \"_acme-challenge.pihole2.2.internal.lan\" \"AVbCsG7_AB0lAzIQDxLBj0c9-gVEntFkaZkfmeoxtLA\";"
[Fri Aug  9 12:49:32 UTC 2024] adding _acme-challenge.pihole2.2.internal.lan. 60 in txt "AVbCsG7_AB0lAzIQDxLBj0c9-gVEntFkaZkfmeoxtLA"
[2024-08-09 12:49:32.69631] [6344] [trace] [bnfHb5pCT8H9] 200 OK (0.177971s, 5.619/s)
[Fri Aug  9 12:49:37 PM UTC 2024] HTTP: 10.2.53.3 POST https://acmeproxy.2.internal.lan:9096/cleanup
[2024-08-09 12:49:37.04000] [6344] [trace] [WpUTr-oUM0Gf] POST "/cleanup"
[2024-08-09 12:49:37.04047] [6344] [trace] [WpUTr-oUM0Gf] Routing to a callback
[Fri Aug  9 12:49:37 PM UTC 2024] auth: acmeuser successfully authenticated for _acme-challenge.pihole2.2.internal.lan.
[Fri Aug  9 12:49:37 PM UTC 2024] executing: /usr/bin/env bash -c "source /root/.acme.sh/acme.sh >/dev/null 2>&1; source /root/.acme.sh/dnsapi/dns_nsupdate.sh; dns_nsupdate_rm \"_acme-challenge.pihole2.2.internal.lan\" \"AVbCsG7_AB0lAzIQDxLBj0c9-gVEntFkaZkfmeoxtLA\";"
[Fri Aug  9 12:49:37 UTC 2024] removing _acme-challenge.pihole2.2.internal.lan. txt
[2024-08-09 12:49:37.11368] [6344] [trace] [WpUTr-oUM0Gf] 200 OK (0.073656s, 13.577/s)


Bind DNS output:

09-Aug-2024 13:47:09.690 update-security: info: client @0x7efe1338c3a0 10.2.53.4#37436/key acme: signer "acme" approved
09-Aug-2024 13:47:09.690 update: info: client @0x7efe1338c3a0 10.2.53.4#37436/key acme: updating zone '2.internal.lan/IN': deleting rrset at '_acme-challenge.pihole2.2.internal.lan' TXT
09-Aug-2024 13:47:09.790 queries: info: client @0x7efe1338c3a0 10.2.53.4#36282/key acme (_acme-challenge.pihole2.2.internal.lan): query: _acme-challenge.pihole2.2.internal.lan IN SOA -S (10.2.53.2)
09-Aug-2024 13:47:09.790 update-security: info: client @0x7efe1338c3a0 10.2.53.4#43179/key acme: signer "acme" approved
09-Aug-2024 13:47:09.790 update: info: client @0x7efe1338c3a0 10.2.53.4#43179/key acme: updating zone '2.internal.lan/IN': adding an RR at '_acme-challenge.pihole2.2.internal.lan' TXT "AVbCsG7_AB0lAzIQDxLBj0c9-gVEntFkaZkfmeoxtLA"


Acme.sh client output:

pihole2:~/.acme.sh# ACMEPROXY_ENDPOINT="https://acmeproxy.2.internal.lan:9096" ACMEPROXY_USERNAME="acmeuser" ACMEPROXY_PASSWORD="********" ./acme.sh --log --issue --dnssleep 3 --
dns dns_acmeproxy -d pihole2.2.internal.lan --keylength ec-384 --cert-file /etc/lighttpd/combined.pem --pre-hook "service lighttpd stop" --reloadcmd "cat /root/.acme.sh/pi
hole2.2.internal.lan_ecc/pihole2.2.internal.lan.key >> /etc/lighttpd/combined.pem && service lighttpd start"
[Fri Aug  9 13:48:10 BST 2024] Using CA: https://ca.internal.lan/acme/stepca/directory
[Fri Aug  9 13:48:10 BST 2024] Runing pre hook:'service lighttpd stop'
 * WARNING: lighttpd is already stopped
[Fri Aug  9 13:48:10 BST 2024] Single domain='pihole2.2.internal.lan'
[Fri Aug  9 13:48:10 BST 2024] Getting webroot for domain='pihole2.2.internal.lan'
[Fri Aug  9 13:48:10 BST 2024] Adding TXT value: AVbCsG7_AB0lAzIQDxLBj0c9-gVEntFkaZkfmeoxtLA for domain: _acme-challenge.pihole2.2.internal.lan
[Fri Aug  9 13:48:10 BST 2024] Using acmeproxy
[Fri Aug  9 13:48:11 BST 2024] Successfully updated the txt record
[Fri Aug  9 13:48:11 BST 2024] The TXT record has been successfully added.
[Fri Aug  9 13:48:11 BST 2024] Sleeping for 3 seconds to wait for the the TXT records to take effect
[Fri Aug  9 13:48:15 BST 2024] Verifying: pihole2.2.internal.lan
[Fri Aug  9 13:48:15 BST 2024] Success
[Fri Aug  9 13:48:15 BST 2024] Removing DNS records.
[Fri Aug  9 13:48:15 BST 2024] Removing txt: AVbCsG7_AB0lAzIQDxLBj0c9-gVEntFkaZkfmeoxtLA for domain: _acme-challenge.pihole2.2.internal.lan
[Fri Aug  9 13:48:15 BST 2024] Using acmeproxy
[Fri Aug  9 13:48:15 BST 2024] Successfully updated the txt record
[Fri Aug  9 13:48:15 BST 2024] Successfully removed
[Fri Aug  9 13:48:15 BST 2024] Verification finished, beginning signing.
[Fri Aug  9 13:48:15 BST 2024] Let's finalize the order.
[Fri Aug  9 13:48:15 BST 2024] Le_OrderFinalize='https://ca.internal.lan/acme/stepca/order/aRMWZKe5ZhSYk7wJYAcYmA4ysogXG9yg/finalize'
[Fri Aug  9 13:48:15 BST 2024] Downloading cert.
[Fri Aug  9 13:48:15 BST 2024] Le_LinkCert='https://ca.internal.lan/acme/stepca/certificate/mE42sAYBCtHyJ4LiDZvOdrOKSIUwph44'
[Fri Aug  9 13:48:15 BST 2024] Cert success.

Success!

My sincere thanks for finding a solution to this, absolutely fantastic.

Best regards,
BAM

**Comment by MadCamel** [comment]
You're welcome. Thanks for sticking around to help test.

I'll add this fix to the latest.
